src/namer.h

```diff
@@ -0,0 +1,129 @@
+#ifndef FLATBUFFERS_NAMER
+#define FLATBUFFERS_NAMER
+
+#include "flatbuffers/idl.h"
+#include "flatbuffers/util.h"
+
+namespace flatbuffers {
+
+// `Namer` applies style configuration to symbols in generated code. It manages
+// casing, escapes keywords, and object API naming.
+// TODO: Refactor all code generators to use this.
+class Namer {
+ public:
+  struct Config {
+    // Symbols in code.
+    Case types;
+    Case constants;
+    Case methods;
+    Case functions;
+    Case fields;
+    Case variants;
+
+    // Namespaces
+    Case namespaces;
+    std::string namespace_seperator;
+
+    // Object API.
+    std::string object_prefix;
+    std::string object_suffix;
+
+    // Keywords.
+    std::string keyword_prefix;
+    std::string keyword_suffix;
+
+    // Files.
+    Case filenames;
+    Case directories;
+    std::string output_path;
+    std::string filename_suffix;
+    std::string filename_extension;
+
+    // DO NOT SUBMIT
+    // comment/doc comment.
+  };
+  Namer(Config config, std::set<std::string> keywords)
+      : config_(config), keywords_(std::move(keywords)) {}
+
+  // Returns a mutable reference to the config so it can be modified **before**
+  // code generation. The intent is for this class to be const while generating
+  // code.
+  Config &GetConfig() { return config_; }
+
+  std::string Type(const std::string &s) const {
+    return Format(s, config_.types);
+  }
+  std::string Method(const std::string &s) const {
+    return Format(s, config_.methods);
+  }
+  std::string Constant(const std::string &s) const {
+    return Format(s, config_.constants);
+  }
+  std::string Function(const std::string &s) const {
+    return Format(s, config_.functions);
+  }
+  std::string Field(const std::string &s) const {
+    return Format(s, config_.fields);
+  }
+  std::string Variant(const std::string &s) const {
+    return Format(s, config_.variants);
+  }
+  std::string ObjectType(const std::string &s) const {
+    return config_.object_prefix + Type(s) + config_.object_suffix;
+  }
+  std::string Namespace(const std::string &s) const {
+    return Format(s, config_.namespaces);
+  }
+  // DO NOT SUBMIT: Add GetRelativeNamespaceTraversal?
+
+  std::string EscapeKeyword(const std::string &name) const {
+    if (keywords_.find(name) == keywords_.end()) {
+      return name;
+    } else {
+      return config_.keyword_prefix + name + config_.keyword_suffix;
+    }
+  }
+  std::string Format(const std::string &s, Case casing) const {
+    // NOTE: If you need to escape keywords after converting case, which would
+    // make more sense than this, make it a config option.
+    return ConvertCase(EscapeKeyword(s), casing, Case::kLowerCamel);
+  }
+  // Returns `filename` with the right casing and suffix and extension.
+  std::string File(const std::string &filename, bool skip_suffix = false,
+                   bool skip_ext = false) const {
+    return ConvertCase(filename, config_.filenames, Case::kUpperCamel) +
```

Comment:
It might be nice to have the expected input type also be its own Config struct? That way we are more readily able to accept non-standard schema definitions?

Reply:
> its own Config struct

I'm not sure what you mean by that but I replaced the booleans with an enum class and I think it's simplified the function a bit. I think we can iterate on how this works as we convert some code generators.

Reply:
I mean for `ConvertCase(filename, config_.filenames, Case::kUpperCamel)` the `Case::kUpperCamel` is hardcoded default of what we expect the schema to be in. It might be nice to put these in a config as well.